pg sources: demux subsource errors

sploiselle opened the pull request:
Previously, an error on any subsource produced an error that prevented all subsources from being used. We can instead de-multiplex subsource errors from the replication stream into _n_ error streams (just as we do for the ok streams). This is a huge UX win as it means one mistake doesn't wedge the entire source.

Additionally, we can include the health of the subsources in the `mz_internal.mz_source_status_history` table.

### Motivation

This PR adds a known-desirable feature. Closes #17490

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Allow reads from all non-errored tables replicated from PostgreSQL into Materialize. Previously, if any table encountered an error during replication, Materialize prohibited reads from all tables replicated via the same source.

philip-stoev commented:
Do you mind if I test this tomorrow? Thank you.

philip-stoev reviewed (approved):
Apologies for the delay and thank you for refactoring the tests. I think they cover all the scenarios that we are relevant.

sploiselle commented on src/storage/src/source/types.rs:
```diff
@@ -213,15 +220,21 @@ pub struct DecodeResult {
+    ///
+    /// This should be 0 to output an error for the primary source; however, you will need to
+    /// determine a numbering scheme to output errors for subsources.
+    pub output: usize,
```
Done

sploiselle commented:
@petrosagg From our discussions:
- Moved the health operator construction into [`build_ingestion_dataflow`](https://github.com/MaterializeInc/materialize/pull/18539/commits/0cd51b606c638a71dc89d74ac5ecad79272ed4c9); tysm for your help with `Scope`
- Addressed https://github.com/MaterializeInc/materialize/pull/18539#discussion_r1163154083 in the last two commits

I left this work in new commits rather than aggressively pushing them down to simplify review, but if there's anything to go back through CI for I'll tidy things up.

petrosagg commented on src/storage/src/render/mod.rs:
```diff
@@ -290,35 +289,46 @@ pub fn build_ingestion_dataflow<A: Allocate>(
                 };
                 tokens.push(token);
+
+                let health_token = crate::source::health_operator(
```
You can delay rendering the health operator until after this for loop. Here you can just collect all the health streams in a vector:

```rust
health_streams.push(health);
```

And then render a single health operator with the combined stream:

```rust
use timely::dataflow::operators::Concatenate;

if let Some(health) = health_streams.pop() {
    let combined_health = health.concatenate(health_streams);
    // A single operator for all the data
    let health_token = crate::source::health_operator(..., combined_health);
}
```

This is desirable since each health operator competes with every other one to compare and append to the shard so it's beneficial to have as few of them as possible. In fact in a future refactor we should make it a single one per `clusterd` but for now we should at least keep their number the same.

sploiselle replied:
ty for this pointer; this actually means we don't need to partition out the stream at all and can just pass it through and then "demux" it in the operator itself by just keeping the `output_index` intact. Makes the operator more complex because it maintains state for each output but I don't think it's _too_ bad. Am going to merge this if it's green and can refactor later if it's too ugly to bear.

petrosagg reviewed (approved):
Looks great! Thanks for pushing this through. I left a comment that I think we should address so that we don't accidentally release something that puts more CRDB pressure but it should be an easy change

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=18539) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=18539) before we can accept your contribution.<br/>**9** out of **10** committers have signed the CLA.<br/><br/>:white_check_mark: def-<br/>:white_check_mark: aalexandrov<br/>:white_check_mark: umanwizard<br/>:white_check_mark: nrainer-materialize<br/>:white_check_mark: tokenrove<br/>:white_check_mark: rkrishn7<br/>:white_check_mark: jkosh44<br/>:white_check_mark: danhhz<br/>:white_check_mark: sploiselle<br/>:x: materializer<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MaterializeInc/materialize?pullRequest=18539) it.</sub>
